lib/promscrape/discovery/azure/api.go

```diff
@@ -0,0 +1,266 @@
+		ActiveDirectoryEndpoint: "https://login.microsoftonline.us",
+		ResourceManagerEndpoint: "https://management.usgovcloudapi.net",
+	}
+)
+
+// apiConfig contains config for API server.
+type apiConfig struct {
+	c              *discoveryutils.Client
+	port           int
+	resourceGroup  string
+	subscriptionID string
+	tenantID       string
+
+	refreshToken tokenRefresher
+	// guards auth token and expiration
+	tokenLock  sync.Mutex
+	authToken  string
+	expiration time.Time
+}
+
+type tokenRefresher func() (string, time.Duration, error)
+
+func getAPIConfig(sdc *SDConfig, baseDir string) (*apiConfig, error) {
+	v, err := configMap.Get(sdc, func() (interface{}, error) { return newAPIConfig(sdc, baseDir) })
+	if err != nil {
+		return nil, err
+	}
+	return v.(*apiConfig), nil
+}
+
+func newAPIConfig(sdc *SDConfig, baseDir string) (*apiConfig, error) {
+	hcc := sdc.HTTPClientConfig
+
+	ac, err := hcc.NewConfig(baseDir)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse auth config: %w", err)
+	}
+	proxyAC, err := sdc.ProxyClientConfig.NewConfig(baseDir)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse proxy auth config: %w", err)
+	}
+
+	cloudEndpoints, err := getCloudEnvByName(sdc.Environment)
+	if err != nil {
+		return nil, err
+	}
+
+	tr, err := newTokenRefresher(sdc, ac, proxyAC, cloudEndpoints)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := discoveryutils.NewClient(cloudEndpoints.ResourceManagerEndpoint, ac, sdc.ProxyURL, proxyAC)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create HTTP client for %q: %w", cloudEndpoints.ResourceManagerEndpoint, err)
+	}
+	cfg := apiConfig{
+		c:              client,
+		subscriptionID: sdc.SubscriptionID,
+		refreshToken:   tr,
+		resourceGroup:  sdc.ResourceGroup,
+		port:           sdc.Port,
+		tenantID:       sdc.TenantID,
+	}
+
+	return &cfg, nil
+}
+
+func getCloudEnvByName(name string) (*cloudEnvironmentEndpoints, error) {
+
+	name = strings.ToUpper(name)
+	// special case, azure cloud k8s cluster, read content from file
+	if name == "AZURESTACKCLOUD" {
+		return readCloudEndpointsFromFile(os.Getenv("AZURE_ENVIRONMENT_FILEPATH"))
+	}
+	env := cloudEnvironments[name]
+	if env == nil {
+		var supportedEnvs []string
+		for envName := range cloudEnvironments {
+			supportedEnvs = append(supportedEnvs, envName)
+		}
+
+		return nil, fmt.Errorf("incorrect value for azure env parametr: %q, supported values: %s", name, strings.Join(supportedEnvs, ","))
+	}
+	return env, nil
+}
+
+func readCloudEndpointsFromFile(filePath string) (*cloudEnvironmentEndpoints, error) {
+	fileContent, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		return nil, fmt.Errorf("cannot read cloud env endpoints from file: %q : %w", filePath, err)
+	}
+	var cee cloudEnvironmentEndpoints
+	if err := json.Unmarshal(fileContent, &cee); err != nil {
+		return nil, fmt.Errorf("cannot parse cloud env endpoints content: %q from file: %q: %w", string(fileContent), filePath, err)
```

Comment:
```suggestion
		return nil, fmt.Errorf("cannot parse cloud env endpoints from file %q: %w", filePath, err)
```
The file content may be too verbose to print... or content sensitive info